src/loader/networks/sdex-chain.ts

```diff
@@ -28,7 +45,182 @@ export class SdexChain {
+      owner,
+      base,
+      quote,
+      poolIdx,
+      lowTick,
+      upperTick,
+    ]);
+
+    const returnData = await this.executeMulticall([{ target: this.config.query, callData }]);
+    const result = this.queryContract.interface.decodeFunctionResult('queryRangeTokens', returnData);
+    return parseRangeTokensResult(result);
+  }
+
+  private validateInputs(owner: string, base: string, quote: string, poolIdx: number) {
+    if (!owner || !base || !quote || poolIdx < 0) {
+      throw new Error('Invalid inputs');
+    }
+  }
+
+  private async getLPTokenBalance(user: string, baseToken: string): Promise<string> {
+    if (baseToken === ZeroAddress) {
+      const balance = await this.context.rpc.getBalance(user);
+      return balance.toString();
+    } else {
+      const LPContract = new ethers.Contract(
+        baseToken,
+        ['function balanceOf(address) view returns (uint256)'],
+        this.context.rpc,
+      );
+      const balance = await LPContract.balanceOf(user);
+      return balance.toString();
+    }
+  }
+
+  private async updateLPTokens(baseToken: string, quoteToken: string) {
+    if (!this.lpTokens.includes(baseToken)) {
+      this.lpTokens.push(baseToken);
+    }
+    if (!this.lpTokens.includes(quoteToken)) {
+      this.lpTokens.push(quoteToken);
+    }
+  }
+
+  public async getUpdatedLiquidity(user: string, base: string, quote: string, poolIdx: number) {
+    const { liquidityChanges } = await this.queryUserPositions(user);
+
+    const concentratedPositions = filterPositions(liquidityChanges, poolIdx, base, quote, PositionType.concentrated);
+    const ambientPositions = filterPositions(liquidityChanges, poolIdx, base, quote, PositionType.ambient);
+
+    if (concentratedPositions.length === 0 && ambientPositions.length === 0) {
+      return null;
+    }
+
+    await this.updateLPTokens(base, quote);
+
+    const ambientPositionResults = await Promise.all(
+      ambientPositions.map(async (userLiquidity) => {
+        const ambientTokens = await this.queryAmbientTokens(user, base, quote, poolIdx);
+        const [lpTokenBalanceBase, lpTokenBalanceQuote] = await Promise.all([
+          this.getLPTokenBalance(user, base),
+          this.getLPTokenBalance(user, quote),
+        ]);
+        const ambientLiq = bignumber(ambientTokens.liq)
+          .plus(bignumber(lpTokenBalanceBase))
+          .plus(bignumber(lpTokenBalanceQuote))
+          .toString();
+
+        return {
+          ambientLiq,
+          concLiq: '0',
+          rewardLiq: '0',
+          baseQty: ambientTokens.baseQty.toString(),
+          quoteQty: ambientTokens.quoteQty.toString(),
+          aggregatedLiquidity: userLiquidity.liq,
+          aggregatedBaseFlow: userLiquidity.baseFlow,
+          aggregatedQuoteFlow: userLiquidity.quoteFlow,
+          positionType: userLiquidity.positionType,
+        };
+      }),
+    );
+
+    const aggregatedAmbientPosition = aggregatePositions(ambientPositionResults);
+
```

Comment:
this would be good place to have multicall:

loop through positions and build multicall data for `queryRangeTokens` and `queryConcRewards` for each of positions.

then use retrieved data to fill in `concentratedPositionsResults` with that data instead of having calls for each position separately in the mapping.